feat: game modifier/customization menu (closes #28)

DeguShi opened the pull request:
## Description
Introduce a proper **Start Menu** and **Settings** screen for existing features.  
Players can choose **Start Game** or **Settings** from the initial screen (mouse/Enter/Space), and can also open Settings mid-game with **M**.  
The Settings UI is data-driven and scrollable, making it easy to add future options.

## Details
- Start screen replaced by a small menu: **Start Game** / **Settings**.  
- Settings (existing knobs only):
  - **Grid size**
  - **Initial speed**
  - **Max speed** (caps progressive increase)
  - **Death sound** (on/off)
- Window + fonts automatically resize when **Grid size** changes.  
- In-game **M** opens Settings and safely reapplies geometry (recreates snake/apple).  
- Game Over screen now **auto-fits text** to window width.  
- Settings control works with both arrow and WASD keys.
- Minor cleanups: removed old `center_prompt`, corrected pause comment (“P”), simplified apple spawn (`randrange` on grid).

## Benefits
- Clear entry point (no “any key” ambiguity) and accessible Settings.  
- Scrollable, declarative menu that new contributors can extend by adding one dict entry.  
- Consistent rendering across resolutions; text always fits.  

## Observations
- This PR also resolves or overlaps with previously mentioned issues related to:
- Gameplay Balance: Maximum Speed Cap is Too High #60 
- feat: fix the sizing of the playable area #19 

@davigdomingues and @FerSuaiden, I ended up modifying parts of the codebase that directly impact your issues.  
There’s still room for refinement, but I believe it’s best to review these updates first before continuing work on your improvements. I apologise for not notifying you in advance, this wasn’t intentional and was done purely in the interest of keeping the implementation consistent and functional.  
Thanks for your understanding and collaboration!

Closes #28 

https://github.com/user-attachments/assets/56d5fcb6-7425-4482-9299-491c7f4106e8

DeguShi committed:
feat(ui): start menu and settings screen

Introduce a proper Start Menu and Settings screen for existing features.
Players can now access Settings before or during gameplay, adjust grid size,
speed values, and toggle death sound. Window and fonts resize automatically
when grid size changes. Game Over text auto-fits to window width.

clr-cera commented:
Hi! Can you change the pr base branch to dev?

clr-cera commented on kobra.py:
```diff
@@ -56,6 +56,196 @@
+
+# Central settings used by the menu; only existing features here.
+SETTINGS = {
+    "grid_size": 50,  # current GRID_SIZE
```
I was confused with the name grid_size, as reducing its value instantiate more cells, what do you think about naming it cell_size or something like that?

DeguShi replied:
Sure! I followed the previous logic as to how the grid size was calculated, but I also got confused at first as to why it was like that, so it is not a problem making it more intuitive. But basically `WIDTH = HEIGHT = (safe_max_dimension // GRID_SIZE) * GRID_SIZE` was saying that, the smaller the grid size, the more cells it would have, which does make sense. Given that, I can proceed in two different ways here, the first one would be to fix the number of cells and only change the grid size (it would affect the window size overall), as for the second one, that would be what already happens, which also changes the number of cells, just adding a more clear naming to it. 

I personally believe that keeping the window size constant is the way to go, so I will follow the second option for now, but I'm open to changing it depending on your feedbacks.

clr-cera replied:
I think the second option is nice!!

DeguShi replied:
<img width="974" height="997" alt="Screenshot 2025-10-10 at 10 20 48" src="https://github.com/user-attachments/assets/a7113994-a996-47e5-9d9c-5c433ea9c657" />
<img width="971" height="996" alt="Screenshot 2025-10-10 at 10 20 41" src="https://github.com/user-attachments/assets/7af60d89-67f9-4959-8357-7fa900651551" />

Perfect! I have made the necessary changes, it is looking like this now!

clr-cera commented on pyproject.toml:
I believe these changes (in pyproject.toml) are not part of this pr, and #69 will fix this, we can wait for #69 to be merged or remove these changes, what do you think?

DeguShi replied:
Well, to be fair, I didn’t change this file directly. The difference probably appeared during the rebase with main or when syncing with dev, so it wasn’t an intentional edit on my part. You can even see from the print that the changes were not made by me.

<img width="1320" height="425" alt="Screenshot 2025-10-10 at 10 46 42" src="https://github.com/user-attachments/assets/f6cf9487-8f42-446e-8c8b-2c375b272506" />

clr-cera replied:
Yess, I know, it appeared when syncing with dev, as it was from a pr made to main and it didnt appear to dev yet.

DeguShi replied:
oh, I get it now. I believe it is indeed correct to first merge #69 as suggested by yourself, so I will wait for it.

DeguShi committed:
feat: add visual refinements

Changes include:
- Adjusted title font size and layout to prevent overlap
- Simplified apple spawning using  on grid
- Minor comment and naming fixes for clarity (grid/cell terminology)

DeguShi committed:
Merge branch 'dev' into feat/28/settings-menu-implementation

DeguShi committed:
fixed build

clr-cera commented:
#69 has been merged! Can you rebase with dev again?

DeguShi committed:
Merge branch 'dev' into feat/28/settings-menu-implementation

DeguShi commented:
@clr-cera it should be good to go now. :)

LiviaLelis commented on .DS_Store:
Can you remove this file and add the `.DS_Store` to the `.gitignore` file?

DeguShi replied:
Good catch, this was a macOS .DS_Store that slipped into the repo. I’ll remove it from the PR and add it to .gitignore so it won’t come back.

DeguShi replied:
Done! :D

DeguShi committed:
repo: ignore macOS .DS_Store and remove from repository

bdogabriel reviewed (changes requested):
Hi! This is a very good addition, nice work! I tested the settings and they seem to work, you just have to rebase to dev and solve any conflicts: `git pull origin dev --rebase`

Also, add this to the description so github closes the issues automatically (I don't know if adding it here in the comments works as well):

Closes #60 
Closes #19

DeguShi committed:
Merge branch 'fossguild:dev' into feat/28/settings-menu-implementation

DeguShi committed:
fix: resolve merge conflicts and finalize settings menu implementation

DeguShi committed:
fix: build problems

DeguShi commented:
@bdogabriel I’ve already selected the issues that will be closed by this PR (see screenshot), so there’s no need to duplicate them in the description.  Regarding the base, I’d already rebased twice earlier, and I’ve just rebased again to ensure the branch is fully up to date with dev. If anything else looks out of sync on your end, let me know and I’ll double-check.

<img width="321" height="145" alt="Screenshot 2025-10-11 at 12 04 54" src="https://github.com/user-attachments/assets/6c95bb33-d713-4d82-bae1-80d6f6309846" />

bdogabriel reviewed (approved):
LGTM

clr-cera reviewed (approved):
LGTM 🥳
